### ISSUE TITLE
Add generics to get/query methods

### ISSUE THREAD

**Issue by nstepien** [testing-library/dom-testing-library#615]
### Describe the feature you'd like:

I know the element type I'm querying, so it'd be beneficial to specify it upfront:
```ts
screen.queryByText<HTMLButtonElement>('click me'); // -> HTMLButtonElement | null
```
Type assertions are very error prone, so should be avoided:
```ts
screen.queryByText('click me') as HTMLButtonElement; // can still be null!
```

### Suggested implementation:

```ts
queryByText<T extends HTMLElement = HTMLElement>(...) => T | null;
```

### Describe alternatives you've considered:

You can do already do something similar with for example `querySelector`:
```ts
document.body.querySelector<HTMLButtonElement>('.my-button');
```

**Comment by kentcdodds** [comment]
Hi @nstepien,

Interestingly, when Testing Library was released, it only supported the `query*` variant. But someone opened an issue about this with TypeScript and we introduced `get*` which can only return the element (otherwise it throws). So while I'm not certain of your specific suggestion (I don't use typescript), I can suggest that you switch to `get*` and avoid the issue you're talking about.

**Comment by nstepien** [comment]
@kentcdodds I might have not made myself clear, sorry.
What I mean is that `queryByText` for example returns `HTMLElement | null` by default, which makes accessing `button.disabled` a type error for example, so I have to assert the type of the returned value.
So my issue is that all the get/query methods return HTMLElement and require me to assert the type, which is error prone.

**Comment by kentcdodds** [comment]
Ah, gotcha. There's currently an effort to migrate this project to TypeScript and maybe when that's finished we can address this. It definitely makes sense.

**Comment by smeijer** [comment]
@kentcdodds, @nstepien is referring purely to the typescript annotations. If I check the ones that are now at definitely typed, they indeed do not support generics.

I'm with @nstepien here, generics would be definitely helpful. It's a way to hint typescript about the return type of the function. Instead of casting it to a given one afterward.

```ts
// hinting
screen.queryByText<HTMLButtonElement>('click me'); // -> HTMLButtonElement | null

// casting
screen.queryByText('click me') as HTMLButtonElement; 
```

The first one can return type errors if the resulting element isn't a `button`. The second will forcefully declare it as being one. Possibly resulting in runtime errors later on.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9fdeede40caaef5e431462f140bec1d3383c0ab9/types/testing-library__dom/queries.d.ts#L68-L86

**Comment by eps1lon** [comment]
Definitely going to strongly advocate for type casting here. More explanation in https://twitter.com/SeaRyanC/status/1090998621472940032

TL;DR: Using generics in this case is just like type casting without the explicitness of type casting which makes it strictly worse.

**Comment by eps1lon** [comment]
> screen.queryByText('click me') as HTMLButtonElement; // can still be null!

That convinced me to go with generics instead. I wish we had a better mechanism because this still masquerades type casting but it's somewhat safer type casting.

**Comment by leomeloxp** [comment]
I think the most appropriate way to supporting generics in here would be to have it extend `HTMLElement` or fallback to it, eg:

```ts
// ...
export type QueryByBoundAttribute<T extends HTMLElement> = (
  container: HTMLElement,
  id: Matcher,
  options?: MatcherOptions,
) => T | null
// ...
export type GetByBoundAttribute<T extends HTMLElement> = (
  container: HTMLElement,
  id: Matcher,
  options?: MatcherOptions,
) => T
// ...
```

**Comment by Daniel-Griffiths** [comment]
I agree with @leomeloxp ‘s solution.

Using type casting allows the user to potentially incorrectly set the types and potentially hit undefined errors at runtime.

Heres an example:

**Using casting**
```
const element = screen.queryByTestId(‘some-id-that-doesnt-exist’) as HTMLImageElement; // Type is HTMLImageElement

doSomething(element.src) // Runtime error, .src does not exist on element
```

**Using generics**
```
const element = screen.queryByTestId<HTMLImageElement>(‘some-id-that-doesnt-exist’); // Type is HTMLImageElement | null

doSomething(element.src) // Error in your code editor telling you to check if element is defined before using it
```

Developers could cast thier code by using `as HTMLImageElement | null` but I think it's easy to forget, and not everyone will know they need to do it. While using generics will enforce the `| null` on the return type and help prevent runtime errors.

**Comment by eps1lon** [comment]
> Using type casting allows the user to potentially set incorrect the types and hit undefined errors at runtime.

So would using generics. Though arguing is moot since all the points you brought up were already included in the initial description.

**Comment by Daniel-Griffiths** [comment]
@eps1lon How about something like this? This would prevent people passing incorrect types, and would also default to `HTMLElement` is no generic types are passed.

```typescript
type Elements = HTMLAnchorElement | HTMLImageElement | HTMLElement; // plus any other HTML types
type ElementsIntial = Elements | { _type: 'initial'};


function getByDataID<T extends ElementsIntial>(
  id: string
): T extends Elements ? T | null : HTMLElement | null {
  // Just some dummy example code so this actually runs
  if (id === "example") {
    return "test" as any;
  }
  return null as any;
}

// Returns HTMLAnchorElement | null
const anchor = getByDataID<HTMLAnchorElement>("test-id");

// Warning: Type 'string' does not satisfy the constraint 'Elements'.
const image = getByDataID<string>("test-id");

// No type specified, returns HTMLElement | null
const element = getByDataID("test-id");

// Object is possibly null
console.log(anchor.href);
```

**Comment by Daniel-Griffiths** [comment]
Or this is another approach, in this example if a user passes the wrong type, we automatically return HTMLElement instead of throwing up a warning.

```
type Elements = HTMLAnchorElement | HTMLImageElement | HTMLElement;

function getByDataID<T>(
  id: string
): T extends Elements ? T | null : HTMLElement | null {
  if (id === "example") {
    return "test" as any;
  }
  return null as any;
}

// Returns HTMLAnchorElement | null
const anchor = getByDataID<HTMLAnchorElement>("test-id");

// Returns HTMLAnchorElement | null
const image = getByDataID<string>("test-id");

// No type specified, returns HTMLElement | null
const element = getByDataID("test-id");

// Object is possibly null
console.log(anchor.href);

```

**Comment by nstepien** [comment]
@Daniel-Griffiths did you read the opening post?

**Comment by Daniel-Griffiths** [comment]
@nstepien I read it again and realised I solved the same problem in a more complicated way, apologies :sweat_smile: 
Although, the code snippet above works a little bit differently where passing in a incorrect type returns HTMLElement by default instead of a warning, although I think a warning would be better in this case. 

I don't want to ping people in this thread more than I already have so I will leave it at that :sweat_smile:

**Comment by nickmccurdy** [comment]
From what I've heard about writing type declarations, using generics only to set the return type without checking or inferring against any other types (such as parameters in an endofunctor) is an antipattern. Casts would be preferred in this case.

**Comment by nstepien** [comment]
I don't know what it is you read about it, but you can't validate that your type assertion extends `HTMLElement` for example, but you can with generics. It's too easy to make mistakes with type assertions, like has been discussed in this thread.

**Comment by timdeschryver** [comment]
With TS 4.1, we could provide type-safety with our `ByRole` queries.
We can infer the element type by using the aria role, something like this:

```ts
export type ExtractElement<
  RoleMatcher extends ByRoleMatcher
> = RoleMatcher extends ARIARole
  ? RoleMatcher extends `button`
    ? HTMLButtonElement
    : RoleMatcher extends `textbox`
    ? HTMLInputElement | HTMLTextAreaElement
    : RoleMatcher extends `checkbox`
    ? HTMLInputElement
    : HTMLElement
  : HTMLElement
```

This seems to work in my POC, but I think my current implementation can be refactored a bit.
If you think this is a good enhancement, I can create a Pull Request and we can discuss it there.

**Comment by eps1lon** [comment]
@timdeschryver  A `role="button"` can be any `Element` not just `HTMLButtonElement`. Same applies to most roles.

**Comment by timdeschryver** [comment]
Right 🤦‍♂️, thanks for the reminder.

**Comment by nstepien** [comment]
![image](https://user-images.githubusercontent.com/567105/133441177-2dd5f9f2-6d8e-4b19-ab16-2e00123517ea.png)
![image](https://user-images.githubusercontent.com/567105/133441341-0448ee30-c9d1-4478-abb7-20d69a2039ff.png)

@eps1lon Can we reopen this?

While the functions are now generic, the way they're set on the `Screen` type erases the generics.

**Comment by eps1lon** [comment]
@nstepien this issue has been closed. If there are other issue with the latest version please open a new one.
